More Python3 fixes

pbauer opened the pull request:
This does not only fix some issues when running in python 3 but also disables all static resources in py3 until that is fixed.

ale-rt commented:
Looks better now...
![image](https://user-images.githubusercontent.com/1300763/39397544-4cab2ff2-4b01-11e8-9bbe-f22e710ee3fc.png)

hvelarde commented on Products/CMFPlone/controlpanel/browser/usergroups_groupmembership.py:
```diff
@@ -92,7 +92,7 @@ def getMembers(self):
 
         mergedResults = groupResults + userResults
-        return filter(None, mergedResults)
+        return [i for i in filter(None, mergedResults)]
```
this is completely wrong, this must be replaced by:

```python
[i for i in mergedResults if i]
```

pbauer replied:
@hvelarde True. You are welcome to fixing it.

hvelarde replied:
@pbauer done! :)

hvelarde commented:
@pbauer how are you dealing with this? are you just replacing things that break something?

we are using Pylint on Python 3 porting mode to help us with our add-ons; something like this would help:

    pylint --py3k --disable=no-absolute-import src/Products/CMFPlone

hvelarde commented:
here's a sample output:

```console
$ pylint --py3k --disable=no-absolute-import Products/CMFPlone/
No config file found, using default configuration
************* Module Products.CMFPlone.PloneFolder
W:274,11: cmp built-in referenced (cmp-builtin)
************* Module Products.CMFPlone.PloneBatch
W: 30, 4: __nonzero__ method defined (nonzero-method)
W: 53,15: map built-in referenced when not iterating (map-builtin-not-iterating)
************* Module Products.CMFPlone.PloneTool
W:1101,44: dict.keys referenced when not iterating (dict-keys-not-iterating)
************* Module Products.CMFPlone.CatalogTool
W:247,15: division w/o __future__ statement (old-division)
W:249,34: division w/o __future__ statement (old-division)
************* Module Products.CMFPlone.MigrationTool
W:218,14: dict.items referenced when not iterating (dict-items-not-iterating)
************* Module Products.CMFPlone.Portal
W:152,36: cmp built-in referenced (cmp-builtin)
************* Module Products.CMFPlone.WorkflowTool
W:248,18: dict.values referenced when not iterating (dict-values-not-iterating)
************* Module Products.CMFPlone.DublinCore
W: 46,15: map built-in referenced when not iterating (map-builtin-not-iterating)
W:390,31: map built-in referenced when not iterating (map-builtin-not-iterating)
************* Module Products.CMFPlone.utils
W:411,26: dict.keys referenced when not iterating (dict-keys-not-iterating)
W:484,29: unicode built-in referenced (unicode-builtin)
W:486,31: basestring built-in referenced (basestring-builtin)
W:488,24: unicode built-in referenced (unicode-builtin)
************* Module Products.CMFPlone.browser.search
W: 57,12: map built-in referenced when not iterating (map-builtin-not-iterating)
W:283,24: dict.keys referenced when not iterating (dict-keys-not-iterating)
W:285,27: dict.keys referenced when not iterating (dict-keys-not-iterating)
************* Module Products.CMFPlone.browser.ploneview
W:122,19: division w/o __future__ statement (old-division)
************* Module Products.CMFPlone._scripts.compile_resources
E: 49, 8: print statement used (print-statement)
E: 55, 4: print statement used (print-statement)
E: 57, 4: print statement used (print-statement)
E:137, 8: print statement used (print-statement)
E:138, 8: print statement used (print-statement)
E:158, 8: print statement used (print-statement)
E:165, 8: print statement used (print-statement)
E:171, 8: print statement used (print-statement)
E:174, 4: print statement used (print-statement)
************* Module Products.CMFPlone._scripts._generate_gruntfile
E: 47, 0: print statement used (print-statement)
E: 52, 0: print statement used (print-statement)
E:200,24: print statement used (print-statement)
E:203,20: print statement used (print-statement)
E:216, 8: print statement used (print-statement)
E:263,12: print statement used (print-statement)
E:306,12: print statement used (print-statement)
E:359,12: print statement used (print-statement)
E:383, 8: print statement used (print-statement)
E:389,12: print statement used (print-statement)
E:390,12: print statement used (print-statement)
E:399,12: print statement used (print-statement)
E:400,12: print statement used (print-statement)
E:441,24: print statement used (print-statement)
************* Module Products.CMFPlone.tests.testResourceRegistries
W:578,12: filter built-in referenced when not iterating (filter-builtin-not-iterating)
W:622,12: filter built-in referenced when not iterating (filter-builtin-not-iterating)
************* Module Products.CMFPlone.tests.testNextPrevious
W: 68,15: Called a next() method on an object (next-method-called)
W: 82,31: Called a next() method on an object (next-method-called)
************* Module Products.CMFPlone.tests.testBatch
W: 10,22: range built-in referenced when not iterating (range-builtin-not-iterating)
W: 17,32: range built-in referenced when not iterating (range-builtin-not-iterating)
************* Module Products.CMFPlone.UnicodeSplitter.config
W: 38, 5: dict.values referenced when not iterating (dict-values-not-iterating)
************* Module Products.CMFPlone.controlpanel.browser.quickinstaller
W:614,25: dict.values referenced when not iterating (dict-values-not-iterating)
W:644,15: dict.values referenced when not iterating (dict-values-not-iterating)
W:647,15: dict.values referenced when not iterating (dict-values-not-iterating)
W:650,15: dict.values referenced when not iterating (dict-values-not-iterating)
W:653,15: dict.values referenced when not iterating (dict-values-not-iterating)
************* Module Products.CMFPlone.controlpanel.browser.usergroups_usermembership
W: 65,15: filter built-in referenced when not iterating (filter-builtin-not-iterating)

-----------------------------------
Your code has been rated at 9.93/10
```

gforcada commented:
@pbauer would it make sense to split this up and merge the safe bits, i.e. what Jenkins already accepts, as to not make this branch unmergeable as it keeps diverting from what master has? :thinking:

pbauer commented:
I'd rather merge the whole thing. This is currently freshly rebased after restapi was merged.

pbauer commented:
### Some py3 changes in CMFPlone are missing!

@jensens @ale-rt @hvelarde @esteele @thet

I'm not sure who or what is to blame but some changes were missing from the python3 branch after I rebased from the current master today.

Here is what happened:

- I saw that the changes to remove p.a.controlpanel need to be rebased into the py3 build to fix buildout there. No complaint there, that is to be expected when working on a branch.
- So I rebased the python3 branch on the current master (same with plone.app.theming). 
- Then buildout and the tests were running agains - great!
- But then about 180 more tests were failing than should and I remember most of these.
- I checked the history and saw that the last change that was left from the python3 branch before the rebase was from 26. Juni (https://github.com/plone/Products.CMFPlone/commit/85714c4500df45b1cccca3bf2d4cdc4dcf7ad141)
- I remember there must have been changes after that. 
- Using the demo-site and the reflog I found the (hopefully) most recent change in the python3 branch https://github.com/plone/Products.CMFPlone/commit/b8599797e12797fd995eee460783547a4ea74ed0

Following the trail from this commit up (using the link to the parent) I found that the following commits were missing (newest to oldest):

https://github.com/plone/Products.CMFPlone/commit/b8599797e12797fd995eee460783547a4ea74ed0
https://github.com/plone/Products.CMFPlone/commit/bd880a9b02c2a6c9013421a6ef5c76acd3162c02
https://github.com/plone/Products.CMFPlone/commit/6399f9937668416bc5324f9614fd74b818cb0e56
https://github.com/plone/Products.CMFPlone/commit/85714c4500df45b1cccca3bf2d4cdc4dcf7ad141
https://github.com/plone/Products.CMFPlone/commit/f417142020415faec79327d26fe427646ae46b2b
https://github.com/plone/Products.CMFPlone/commit/078d978de35abe9482443d5aab0940dbf84b82e7
https://github.com/plone/Products.CMFPlone/commit/ec98635e0bb9d86a74ddd5be64b6236caf67543d
https://github.com/plone/Products.CMFPlone/commit/88e56ff73718dcdfdc871a956e284bcc81dd3a39
https://github.com/plone/Products.CMFPlone/commit/f776737540bfe357ee2fda23d5b22f0d71ad9976
https://github.com/plone/Products.CMFPlone/commit/db6fd75560f27d53213b4fbafe8012039d446cd6
https://github.com/plone/Products.CMFPlone/commit/97511aedc84a7f9675e98d256d87329b3a6592bf
https://github.com/plone/Products.CMFPlone/commit/5321dc5d94c0d2a814684754c59ff362672ab20c
https://github.com/plone/Products.CMFPlone/commit/b699d94970100623d4fe996408833d42e3e56337
https://github.com/plone/Products.CMFPlone/commit/65f6ae2eee9c03e3014a6f88661136137c929738
https://github.com/plone/Products.CMFPlone/commit/0cc3c0e426d5f122d290547cfac89dae53a24ec4
https://github.com/plone/Products.CMFPlone/commit/11482a2d37b9ded04d517bd1e9f50e234e3b3b46
https://github.com/plone/Products.CMFPlone/commit/61b5959ed826e6a21c5127a86d6a890564556362
https://github.com/plone/Products.CMFPlone/commit/53d31d94949de78a4a4295f9dea1f7ee77c92c62

- I checkout the newest commit and create a new branch (fixup_py3) 
- I created a merge-request for this into python3: https://github.com/plone/Products.CMFPlone/pull/2499/
- I finally realized that the issue with the 180 more failing tests is actually way older than today. It started on August 8: https://jenkins.plone.org/view/PLIPs/job/plip-py3/628
- Since I did not do any changes between 07.08.2018 18:59:00 and 08.08.2018 06:59:00 I feel somewhat vindicated but I could not find out who broke that.

Takeaways:
- git reflog rules!
- Be careful when doing rebases and force-pushes, your local changes might be outdated and the git pull you did might have aborted with a message you ignored (just a wild guess since that happend to me more than once)
- After doing changes in the python3-branches check that https://jenkins.plone.org/view/PLIPs/job/plip-py3 is not totally botched (it runs twice day)

I plan to rebase/review/merge most of the open py3 pull-request during the course of the next week. I take a week-long sailing course and will have some time in the evening.

Sorry for the noise,
Philip

jensens commented:
It could me to blame. Not sure so. A lesson learned from the trouble: Rebase early, rebase often.

pbauer commented:
... and after a merge we're back to 271 test-failures.

hvelarde commented:
@pbauer this pull request is a pain in the but, it currently includes 145 commits and modifies 122 files.

I found this approach complex; Python 3 fixes may be included in batches and that will make our life easier: first push the fixes that are obvious and don't break anything in Python 2.7; this way we can test those fixes earlier for issues. after that we can concentrate on fixing the failing tests.

yes, I know that this can be hard at this point, but I'm just suggesting it.

ale-rt commented on Products/CMFPlone/tests/testPloneTool.py:
```diff
@@ -369,13 +369,21 @@ def testSetFormat(self):
+        # DefaultReadFile.mimeType and allways returns text/plain
+        # for zero or one primary fields
+        # TODO: This seems like a bug.
+        self.assertEqual(self.doc.Format(), 'text/plain')
```
@pbauer Theoretically this one fixes it https://github.com/plone/plone.dexterity/pull/87 but fails in other places.
I tried a more backward compatible PR https://github.com/plone/plone.dexterity/pull/88 but it is also failing
Probably thios is something that we can solve at the sprint
